### PR TITLE
DEV-7247: Added support for np.datetime64 and custom dates

### PR DESCRIPTION
**Pull request by thomasdotred**
DEV-7347: Add np.datetime64 to DateOrCutLabel
DEV-7347: Added custom date format support
DEV-7347: Added custom date format support
DEV-7347: Update Docstrings and test names

# Pull Request Checklist

- [ ] Read the [contributing guidelines](https://github.com/finbourne/lusid-python-tools/blob/master/docs/CONTRIBUTING.md)
- [ ] Tests pass

# Description of the PR

1) Included support for numpy.datetime64  in `DateOrCurLabel`

2) Also added support for custom date formats by passing in the `custom_date_format` e.g `"%Y-%m-%d %H:%M:%S.%f"` . Note that load_from_date_frame still needs to be refactored to pass the custom_date_format into and calls to `DateOrCutLable`. --> DEV-7393

### CODE REVIEW

File: lusidtools/cocoon/dateorcutlabel.py [hunk 1]
```diff
@@ -8,14 +8,33 @@
 
 
 class DateOrCutLabel(UserString):
-    def __init__(self, datetime_value):
-        def convert_datetime_utc(datetime_value):
+    def __init__(self, datetime_value, custom_date_format=None):
```
Comment (rizwansaeed) [inline lusidtools/cocoon/dateorcutlabel.py]: I'd rename to just `date_format`

File: lusidtools/cocoon/dateorcutlabel.py [hunk 2]
```diff
@@ -8,14 +8,33 @@
...
+        def convert_datetime_utc(datetime_value, custom_date_format=None):
             """
             This function ensures that a variable is a timezone aware UTC datetime
 
             :param any datetime_value:
```
Comment (rizwansaeed) [inline lusidtools/cocoon/dateorcutlabel.py]: missing docstring for parameter

File: lusidtools/cocoon/dateorcutlabel.py [hunk 3]
```diff
@@ -8,14 +8,33 @@
...
             :return: datetime datetime_value: The converted timezone aware datetime in the UTC timezone
             """
+            if custom_date_format:
+                if not isinstance(datetime_value, str):
+                    raise TypeError(f"Date {datetime_value} is of type {type(datetime_value)} must be of type 'str' "
+                                    f"when specifying a custom date format. ")
+
+                try:
+                    datetime_value = pd.to_datetime(datetime_value, format=custom_date_format,
```
Comment (rizwansaeed) [inline lusidtools/cocoon/dateorcutlabel.py]: shouldn't this be `utc=True`?

Comment (rizwansaeed) [inline lusidtools/cocoon/dateorcutlabel.py]: sonar might complain about setting a value to a parameter

File: lusidtools/cocoon/dateorcutlabel.py [hunk 4]
```diff
@@ -8,14 +8,33 @@
...
+                                                    utc=False)
+                except ValueError:
+                        raise ValueError(f"The date format provided {custom_date_format} was not recognised in the"
+                                         f" datetime provided: {datetime_value}")
+
+                def formatdt(dt):
```
Comment (rizwansaeed) [inline lusidtools/cocoon/dateorcutlabel.py]: rename to `format_dt`

File: lusidtools/cocoon/dateorcutlabel.py [hunk 5]
```diff
@@ -66,6 +85,17 @@ def convert_datetime_utc(datetime_value):
...
 
+
+            # If datetime is numpy datetime convert to ISO format and parse to string
+            if isinstance(datetime_value, np.datetime64):
+                datetime_value = str(
+                    np.datetime_as_string(
+                        arr=datetime_value, timezone="UTC", unit="us"
+                    )
+                )
+                print(datetime_value)
```
Comment (rizwansaeed) [inline lusidtools/cocoon/dateorcutlabel.py]: remove debug

File: tests/unit/cocoon/test_dateorcutlabel.py [hunk 6]
```diff
@@ -69,21 +69,51 @@ def setUpClass(cls) -> None:
...
+                "2019-09-01T06:30:30.005001Z",
+            ],
         ]
     )
     def test_dateorcutlabel(self, test_name, datetime_value, expected_outcome):
         ignore = ["A date with month first"]
         if test_name in ignore:
             self.skipTest("Test not implemented ")
-        date_or_cut_label = DateOrCutLabel(datetime_value)
+        if "custom date" in test_name:
```
Comment (rizwansaeed) [inline tests/unit/cocoon/test_dateorcutlabel.py]: make these different tests, try not to have if logic in tests

File: lusidtools/cocoon/dateorcutlabel.py (hunk 1 above)
Comment (thomasdotred) [reply lusidtools/cocoon/dateorcutlabel.py]: renamed

File: lusidtools/cocoon/dateorcutlabel.py (hunk 2 above)
Comment (thomasdotred) [reply lusidtools/cocoon/dateorcutlabel.py]:
```python
            """
            This function ensures that a variable is a timezone aware UTC datetime

            :param any datetime_value: Datetime variable
            :param any date_format:  The format of a custom date as a string eg "%Y-%m-%d %H:%M:%S.%f". see https://strftime.org/
            :return: datetime datetime_value: The converted timezone aware datetime in the UTC timezone
            """
```

File: lusidtools/cocoon/dateorcutlabel.py (hunk 3 above)
Comment (thomasdotred) [reply lusidtools/cocoon/dateorcutlabel.py]: UTC is irrelevant at this stage as it is set later on line 37. I have removed the argument altogether to reduce conduction.
 
Datetime value is passed into `formatdt()` which extracts "%Y-%m-%d %H:%M:%S.%f" and then formats it to UTC
```
np.datetime_as_string(
                                        arr=np.datetime64(dt.strftime("%Y-%m-%d %H:%M:%S.%f")), timezone="UTC", unit="us"
                                    )
```

Comment (thomasdotred) [reply lusidtools/cocoon/dateorcutlabel.py]: Why is that a bad thing?

**Pull request closed without merging**